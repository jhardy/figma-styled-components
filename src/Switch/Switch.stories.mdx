import { Switch } from "./Switch";
import { Story, Meta } from "@storybook/addon-docs/blocks";

<Meta title="MDX|Switch" component={Switch} />

# Switch

<Story name="Switch">
  <div class="component-box">
    <Switch label="I am a Switch click me" />
  </div>
</Story>

<div class="props-table">

## Props

| Prop Name | Type     | Default | Required | Description                              |
| :-------- | :------- | :-----: | :------: | :----------------------------------------|
| label     | string   |    -    |   yes    | The Switch's label                     |
| checked   | boolean  |  false  |    -     | If the Switch is checked or not        |
| onChange  | function |    -    |    -     | Callback for Switch's `onChange` event |

</div>
