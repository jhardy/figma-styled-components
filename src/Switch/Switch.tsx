import styled from "styled-components";
import * as React from "react";

import {Text} from '../Text'

const SwitchSlider = styled.span`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  transition: transform 0.2s;
  transition: background-color 0 0.2s;
  border: 1px solid #000000;
  border-radius: 12px;
  background-color: #ffffff;

  &:before {
    position: absolute;
    top: -1px;
    left: -1px;
    width: 10px;
    height: 10px;
    content: "";
    transition: transform 0.2s;
    transition: background-color 0 0.2s;
    border: 1px solid #000000;
    border-radius: 50%;
    background-color: white;
  }
`;

export interface SwitchContainerProps {
  checked?: boolean;
  onChange: () => void;
}

const SwitchContainerFactory: React.FC<SwitchContainerProps> = ({ onChange, checked, ...props }) => {
  return (
    <div {...props}>
      <input type="checkbox" onChange={onChange} checked={checked} />
      <SwitchSlider />
    </div>
  );
};

const SwitchContainer = styled(SwitchContainerFactory)`
  position: relative;
  width: 24px;
  height: 12px;
  margin: 10px 16px 10px 8px;

  input {
    width: 0;
    height: 0;
    opacity: 0;
  }

  input:checked + ${SwitchSlider} {
    background-color: #000000;
  }

  input:focus {
    outline: none;
    box-shadow: 0 0 1px #2196f3;
  }

  input:checked + ${SwitchSlider}:before {
    transform: translateX(12px);
  }
`;

export interface SwitchProps {
  onChange?: () => void
  checked?: boolean
  label: string
}
const SwitchFactory: React.FC<SwitchProps> = ({ checked, onChange, label, ...props }) => {
  return (
    <label {...props}>
      <SwitchContainer onChange={() => {onChange}} checked={checked} />
      <Text>{label}</Text>
    </label>
  );
};

export const Switch = styled(SwitchFactory)`
  position: relative;
  display: flex;
  align-items: center;
  align-self: 1;
  flex-direction: row;
  cursor: default;
`;
