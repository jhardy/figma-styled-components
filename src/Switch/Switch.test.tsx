import 'jest-styled-components'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { Switch } from './Switch'


test('A default Switch', () => {
  const tree = renderer.create(<Switch label='A Switch' value='default' />).toJSON()
  expect(tree).toMatchSnapshot()
})


test('A checked Switch', () => {
  const tree = renderer.create(<Switch label='A checked Switch' checked={true} value='checked' />).toJSON()
  expect(tree).toMatchSnapshot()
})


