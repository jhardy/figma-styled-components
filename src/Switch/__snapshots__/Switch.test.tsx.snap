// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A checked Switch 1`] = `
.c4 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 400;
  font-size: 11px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  line-height: 16px;
}

.c3 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transition: background-color 0 0.2s;
  transition: background-color 0 0.2s;
  border: 1px solid #000000;
  border-radius: 12px;
  background-color: #ffffff;
}

.c3:before {
  position: absolute;
  top: -1px;
  left: -1px;
  width: 10px;
  height: 10px;
  content: "";
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transition: background-color 0 0.2s;
  transition: background-color 0 0.2s;
  border: 1px solid #000000;
  border-radius: 50%;
  background-color: white;
}

.c1 {
  position: relative;
  width: 24px;
  height: 12px;
  margin: 10px 16px 10px 8px;
}

.c1 input {
  width: 0;
  height: 0;
  opacity: 0;
}

.c1 input:checked + .c2 {
  background-color: #000000;
}

.c1 input:focus {
  outline: none;
  box-shadow: 0 0 1px #2196f3;
}

.c1 input:checked + .c2:before {
  -webkit-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: 1;
  -ms-flex-item-align: 1;
  align-self: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  cursor: default;
}

<label
  className="c0"
>
  <div
    className="c1"
  >
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="c2 c3"
    />
  </div>
  <span
    className="c4"
  >
    A checked Switch
  </span>
</label>
`;

exports[`A default Switch 1`] = `
.c4 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 400;
  font-size: 11px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  line-height: 16px;
}

.c3 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transition: background-color 0 0.2s;
  transition: background-color 0 0.2s;
  border: 1px solid #000000;
  border-radius: 12px;
  background-color: #ffffff;
}

.c3:before {
  position: absolute;
  top: -1px;
  left: -1px;
  width: 10px;
  height: 10px;
  content: "";
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transition: background-color 0 0.2s;
  transition: background-color 0 0.2s;
  border: 1px solid #000000;
  border-radius: 50%;
  background-color: white;
}

.c1 {
  position: relative;
  width: 24px;
  height: 12px;
  margin: 10px 16px 10px 8px;
}

.c1 input {
  width: 0;
  height: 0;
  opacity: 0;
}

.c1 input:checked + .c2 {
  background-color: #000000;
}

.c1 input:focus {
  outline: none;
  box-shadow: 0 0 1px #2196f3;
}

.c1 input:checked + .c2:before {
  -webkit-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: 1;
  -ms-flex-item-align: 1;
  align-self: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  cursor: default;
}

<label
  className="c0"
>
  <div
    className="c1"
  >
    <input
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="c2 c3"
    />
  </div>
  <span
    className="c4"
  >
    A Switch
  </span>
</label>
`;
